libobs/audio-monitoring/nix/pulseaudio-output.c

```diff
@@ -0,0 +1,487 @@
+
+		pulseaudio_lock();
+		pa_stream_write(data->stream, buffer, bytesToFill, NULL,
+		                      0LL,
+		                      PA_SEEK_RELATIVE);
+		pulseaudio_unlock();
+
+		data->bytesRemaining -= bytesToFill;
+	}
+}
+
+static void on_audio_playback(void *param, obs_source_t *source,
+                              const struct audio_data *audio_data, bool muted)
+{
+	struct audio_monitor *monitor = param;
+	float vol = source->user_volume;
+	size_t bytes;
+
+	uint8_t *resample_data[MAX_AV_PLANES];
+	uint32_t resample_frames;
+	uint64_t ts_offset;
+	bool success;
+
+	if (pthread_mutex_trylock(&monitor->playback_mutex) != 0) {
+		return;
+	}
+	if (os_atomic_load_long(&source->activate_refs) == 0) {
+		goto unlock;
+	}
+
+	success = audio_resampler_resample(monitor->resampler, resample_data,
+	                                   &resample_frames, &ts_offset,
+	                                   (const uint8_t *const *) audio_data->data,
+	                                   (uint32_t) audio_data->frames);
+	if (!success) {
+		goto unlock;
+	}
+
+	bytes = monitor->bytes_per_frame * resample_frames;
+
+	if (muted) {
+		memset(resample_data[0], 0, bytes);
+	} else {
+		if (!close_float(vol, 1.0f, EPSILON)) {
+			process_volume(monitor, vol, resample_data,
+			               resample_frames);
+		}
+	}
+
+	circlebuf_push_back(&monitor->new_data, resample_data[0], bytes);
+	monitor->packets++;
+	monitor->frames += resample_frames;
+
+unlock:
+	pthread_mutex_unlock(&monitor->playback_mutex);
+	do_stream_write(param);
+}
+
+static void
+pulseaudio_stream_write(pa_stream *p, size_t nbytes, void *userdata)
+{
+	UNUSED_PARAMETER(p);
+	PULSE_DATA(userdata);
+
+	pthread_mutex_lock(&data->playback_mutex);
+	data->bytesRemaining += nbytes;
+	pthread_mutex_unlock(&data->playback_mutex);
+
+	pulseaudio_signal(0);
+}
+
+static void pulseaudio_server_info(pa_context *c, const pa_server_info *i,
+                                   void *userdata)
+{
+	UNUSED_PARAMETER(c);
+	UNUSED_PARAMETER(userdata);
+
+	blog(LOG_INFO, "Server name: '%s %s'",
+	     i->server_name, i->server_version);
+
+	pulseaudio_signal(0);
+}
+
+static void
+pulseaudio_source_info(pa_context *c, const pa_source_info *i, int eol,
+                       void *userdata)
+{
+	UNUSED_PARAMETER(c);
+	PULSE_DATA(userdata);
+	// An error occured
+	if (eol < 0) {
+		data->format = PA_SAMPLE_INVALID;
+		goto skip;
+	}
+	// Terminating call for multi instance callbacks
+	if (eol > 0)
+		goto skip;
+
+	blog(LOG_INFO, "Audio format: %s, %"
+		PRIu32
+		" Hz"
+			", %"
+		PRIu8
+		" channels",
+	     pa_sample_format_to_string(i->sample_spec.format),
+	     i->sample_spec.rate,
+	     i->sample_spec.channels);
+
+	pa_sample_format_t format = i->sample_spec.format;
+	if (pulseaudio_to_obs_audio_format(format) == AUDIO_FORMAT_UNKNOWN) {
+		format = PA_SAMPLE_S16LE;
+
+		blog(LOG_INFO, "Sample format %s not supported by OBS,"
+			"using %s instead for recording",
```

Comment:
Style:  Indent with two tabs to continue a line.  See also, the next two lines in this `blog` call.